### ISSUE TITLE
Refinement via disjunctive normal form.

### ISSUE THREAD

**Issue by FrankBro** [josefs/Gradualizer#256]
Hey guys. Was about to finish #245 when I got sidetracked a little bit. It seems like the refinement process when threading through the different function clauses is not as exhaustive as it could be.

Here is an example that works:

```erlang
-spec test(integer() | undefined, integer()) -> integer().
test(undefined, I) -> I;
test(I, _) -> I.
```

Via the refinement process, I'll write down what sort of happens. I'm just free-writing this so it might not be perfect but you'll get the idea.

We enter `check_clause` for the first clause. Here:
 
```erlang
Args :: [{atom, _, undefined}, {var, _, 'I'}]

ArgsTy :: [
    {type, _, union, [
        {atom, _, undefined}, 
        {type, _, integer, []}
    ]}, 
    {type, _, integer, []}
]
```

Via `add_types_pats`, we will extract the type of the pattern. Here:

```erlang
PatTys :: [{atom, _, undefined, {type, _, integer, []}]
```

With the `refine_clause_arg_tys`, we will be able to refine the first argument. We've matched on `{undefined, integer()}`. On the next call:

```erlang
ArgsTy :: [{type, _, integer, []}, {type, _, integer, []}]
```

Great!

Here is where we will have a problem:

```erlang
-spec test(integer() | undefined, integer() | undefined) -> integer().
test(undefined, undefined) -> 0;
test(A, undefined) -> A;
test(undefined, B) -> B;
test(A, B) -> A + B.
```

Uhoh. We get an error here:

```
The variable on line 5 at column 23 is expected to have type integer() but it has type integer() | undefined

-spec test(undefined | integer(), undefined | integer()) -> integer().
test(undefined, undefined) -> 0;
test(A, undefined) -> A;
                      ^
```

If instead of two arguments, we have a tuple. It works.

```erlang
-spec test({integer() | undefined, integer() | undefined}) -> integer().
test({undefined, undefined}) -> 0;
test({A, undefined}) -> A;
test({undefined, B}) -> B;
test({A, B}) -> A + B.
```

What seems to be the problem is that, while the tuple version is allowed to be expanded to all possibilities, the multiple args isn't. We enclose function args in a tuple and treat it this way but we don't allow function args to exit in any other form than a tuple. Actually, the refine worked. If we check here https://github.com/josefs/Gradualizer/blob/master/src/typechecker.erl#L3282 the type it extracted after the first call is:

```erlang
{type,0,union,
    [{type,0,tuple,
            [{type,0,integer,[]},
            {type,{3,35},
                    union,
                    [{type,{3,35},integer,[]},
                    {atom,{3,47},undefined}]}]},
        {type,0,tuple,
            [{type,{3,12},
                    union,
                    [{type,{3,12},integer,[]},
                    {atom,{3,24},undefined}]},
            {type,0,integer,[]}]}]}
```

My idea to fix this was to, along with keeping the `ArgsTy`, keeping a full function refined type. We'll also need to write a function to collapse it back to the initial signature.

```erlang
%% Collapsed
(a | b, c | d)
%% Expanded
(a, c) | (a, d) | (b, c) | (b, d)
```

Does that make sense?

**Comment by zuiderkwast** [comment]
I understand.

Just one concern. Fully expanding like this gives an explosion in the size of the union. Consider expanding a type like `{1..10, 1..10, 1..10}`. It expands to a union with 10<sup>3</sup> elements. Perhaps it can be mitigated using lazy unions, but it gets more complex. (See https://www.irif.fr/~gc/papers/covcon-again.pdf)

**Comment by FrankBro** [comment]
Good point. I saw somewhere there's a performance hack to not evaluate a tuple over 20 elements. Also, I did not think about integers. What if we skip em?

**Comment by zuiderkwast** [comment]
It could work, I guess. I don't like the hacks such as the one limiting to 20 elements though. We should consider it temporary. Also the refinement rules shouldn't be too complex. It should be documented and easy to predict for the users.

**Comment by FrankBro** [comment]
I will like the example I wrote is pretty predictable for a user. I was actually surprised it didn't work.

**Comment by zuiderkwast** [comment]
Sure.

The main reason this doesn't already work is that a clause has multiple args and types and we give type errors for one type at a time.

So, do you want the tuple-wrapping to happen earlier, so that in `check_clause`, the argument is a single tuple and the expected type is a single tuple type?

**Comment by FrankBro** [comment]
The way I was trying to implement it was via initially generating the full set of possibilities and keeping it along while we do every individual clauses. Also, I'd need a function to collapse back all the possibilities into a tuple of args. We have a way to go A to B but not sure we have a way to go B to A.

```erlang
      A                            B
(a | b, c | d) <=> (a, c) | (a, d) | (b, c) | (b, d)
```

**Comment by josefs** [comment]
@FrankBro it's good to see that you're thinking about these things. The combination of unions and tuples is a problem for gradualizer because they cause a performance problem. That's why we currently struggle to applying gradualizer to itself. To me that's a more pressing problem than getting refinement right and I hope to start looking at the performance problem soon.

**Comment by zuiderkwast** [comment]
DNF refers to the logic AND/OR/NOT (or the set-theoretic counterparts): https://en.wikipedia.org/wiki/Disjunctive_normal_form, so it's perhaps not the right term for expanding tuples-unions. Anyway, we get a lot of unions here too. :-)

A way to limit the explosion could be by adding lazy unions.

We could add a "type" on the form

```Erlang
{type, _, lazy_union, NormalizedTypes, NotYetNormalizedTypes}
```

which is only partially normalized. When we check subtyping and other things, we try the NormalizedTypes part first and only process the lazy part when necessary.

I haven't thought about exactly when/where/how it should be created and used, but maybe it could work. What do you think?

**Comment by FrankBro** [comment]
I'm not sure how much it would help but this is mostly due to the fact I haven't looked into the performance bottleneck yet. I was thinking of running fprof once I finished my record refinement branch. If that's the case, I guess the logical thing to do would be to split this work in it's own PR.
